src/ehrdata/_feature_types.py

```diff
@@ -0,0 +1,326 @@
+    """Detect the feature type of a column in a pandas DataFrame.
+
+    Args:
+        col: The column to detect the feature type for.
+
+    Returns:
+        The detected feature type (one of 'date', 'categorical', or 'numeric') and a boolean, which is True if the feature type is uncertain.
+    """
+    n_elements = len(col)
+    col = col.replace(MISSING_VALUES, np.nan)
+    col = col.dropna()
+    if len(col) == 0:
+        err_msg = f"Feature '{col.name}' contains only NaN values. Please drop this feature to infer the feature type."
+        raise ValueError(err_msg)
+    majority_type = col.apply(type).value_counts().idxmax()
+
+    if majority_type == pd.Timestamp:
+        return DATE_TAG, False  # type: ignore
+
+    if majority_type is str:
+        try:
+            col.apply(isoparse)
+            return DATE_TAG, False  # type: ignore
+        except ValueError:
+            try:
+                col = pd.to_numeric(col, errors="raise")  # Could be an encoded categorical or a numeric feature
+                majority_type = float
+            except ValueError:
+                # Features stored as Strings that cannot be converted to float are assumed to be categorical
+                return CATEGORICAL_TAG, False  # type: ignore
+
+    if majority_type not in [int, float]:
+        return CATEGORICAL_TAG, False  # type: ignore
+
+    # Guess categorical if the feature is an integer and the values are 0/1 to n-1/n with no gaps
+    if (
+        (majority_type is int or (np.all(i.is_integer() for i in col)))
+        and (n_elements != col.nunique())
+        and (
+            (col.min() == 0 and np.all(np.sort(col.unique()) == np.arange(col.nunique())))
+            or (col.min() == 1 and np.all(np.sort(col.unique()) == np.arange(1, col.nunique() + 1)))
+        )
+    ):
+        return CATEGORICAL_TAG, True  # type: ignore
+
+    return NUMERIC_TAG, False  # type: ignore
+
+
+def infer_feature_types(
+    edata: EHRData,
+    *,
+    layer: str | None = None,
+    output: Literal["tree", "dataframe"] | None = "tree",
+    verbose: bool = True,
+) -> pd.DataFrame | None:
+    """Infer feature types from an :class:`~ehrdata.EHRData` object.
+
+    For each feature in `edata.var_names`, the method infers one of the following types: `'date'`, `'categorical'`, or `'numeric'`.
+    The inferred types are stored in `edata.var['feature_type']`. Please check the inferred types and adjust if necessary using
+    `edata.var['feature_type']['feature1']='corrected_type'` or with :func:`~ehrdata.replace_feature_types`.
+    Be aware that not all features stored numerically are of `'numeric'` type, as categorical features might be stored in a numerically encoded format.
+    For example, a feature with values [0, 1, 2] might be a categorical feature with three categories. This is accounted for in the method, but it is
+    recommended to check the inferred types.
+
+    Args:
+        edata: Data object.
```

Comment:
I wonder whether we can have something more descriptive than that. I like the AnnData version.